OptimoveSDK/Classes/NotificationHandler/OptimoveNotificationHandler.swift

```diff
@@ -104,55 +105,60 @@ private extension OptimoveNotificationHandler {
-            Logger.warn("Campaign details could not be extracted. Stop execution.")
-            return
-        }
-
-        let task = UIApplication.shared.beginBackgroundTask(withName: "notification reponse")
-        switch response.actionIdentifier {
-        case UNNotificationDismissActionIdentifier:
-            optimove.reportEvent(NotificationDismissedEvent(campaignDetails: campaignDetails))
-            optimove.dispatchQueuedEventsNow()
-            DispatchQueue.main.asyncAfter(deadline: .now() + min(UIApplication.shared.backgroundTimeRemaining, 2.0)) {
-                UIApplication.shared.endBackgroundTask(task)
-            }
-
-        case UNNotificationDefaultActionIdentifier:
-            optimove.reportEvent(NotificationOpenedEvent(campaignDetails: campaignDetails))
-            optimove.dispatchQueuedEventsNow()
-            DispatchQueue.main.asyncAfter(deadline: .now() + min(UIApplication.shared.backgroundTimeRemaining, 2.0)) {
+        do {
+            let event = try createEvent(from: response)
+            let task = UIApplication.shared.beginBackgroundTask(withName: "Handling a notification reponse")
+            switch response.actionIdentifier {
+            case UNNotificationDefaultActionIdentifier:
+                optimove.reportEvent(event)
+                optimove.dispatchQueuedEventsNow()
+                let delay: TimeInterval = min(UIApplication.shared.backgroundTimeRemaining, 2.0)
+                DispatchQueue.global().asyncAfter(deadline: .now() + delay) {
+                    UIApplication.shared.endBackgroundTask(task)
+                }
+            default:
                 UIApplication.shared.endBackgroundTask(task)
             }
+        } catch {
+            Logger.error(error.localizedDescription)
+        }
+    }
 
-        default: UIApplication.shared.endBackgroundTask(task)
+    func createEvent(from response: UNNotificationResponse) throws -> OptimoveCoreEvent {
+        let notificationDetails = response.notification.request.content.userInfo
+        let data = try JSONSerialization.data(withJSONObject: notificationDetails)
+        let notification = try JSONDecoder().decode(NotificationPayload.self, from: data)
+        switch notification.campaign {
+        case let campaign as ScheduledNotificationCampaign:
+            return ScheduledNotificationOpened(
+                bundleIdentifier: try Bundle.getApplicationNameSpace(),
+                campaign: campaign
+            )
+        case let campaign as TriggeredNotificationCampaign:
+            return TriggeredNotificationOpened(
+                bundleIdentifier: try Bundle.getApplicationNameSpace(),
+                campaign: campaign
+            )
+        default:
+            throw GuardError.custom("The campaign of type \(notification.campaign.type.rawValue) is not supported.")
         }
     }
 
     func handleDeepLinkDelegation(_ response: UNNotificationResponse) {
```

Comment:
@eligoptimove What are the test cases for this method?

Reply:
Missed. Will add asap.

Reply:
There is needs to provide more complex testing and refactoring than just a test. Will create another branch for this task.